test/e2e/json-rpc/wallet_addEthereumChain.spec.ts

```diff
@@ -0,0 +1,329 @@
+          );
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+          await driver.clickElement({ text: 'Confirm', tag: 'button' });
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+          const afterPermittedChains = await getPermittedChains(driver);
+          assert.deepEqual(afterPermittedChains, ['0x53a']);
+        },
+      );
+    });
+  });
+
+  describe('the dapp is already permitted to use the chain being added', () => {
+    it('automatically switches to the chain when the rpc endpoint is added but a different rpc endpoint already existed for the chain', async function () {
+      await withFixtures(
+        {
+          dapp: true,
+          fixtures: new FixtureBuilder()
+            .withNetworkControllerDoubleGanache()
+            .withPermissionControllerConnectedToTestDappWithChains([
+              '0x539',
+              '0x53a',
+            ])
+            .build(),
+          ganacheOptions: {
+            ...defaultGanacheOptions,
+            concurrent: [{ port: 8546, chainId: 1338 }],
+          },
+          title: this.test?.fullTitle(),
+        },
+        async ({ driver }: { driver: Driver }) => {
+          await unlockWallet(driver);
+          await openDapp(driver);
+
+          const beforePermittedChains = await getPermittedChains(driver);
+          assert.deepEqual(beforePermittedChains, ['0x539', '0x53a']);
+
+          // should start on 1337
+          await driver.findElement({ css: '#chainId', text: '0x539' });
+
+          const switchEthereumChainRequest = JSON.stringify({
+            jsonrpc: '2.0',
+            method: 'wallet_addEthereumChain',
+            params: [
+              {
+                chainId: '0x53a',
+                chainName: 'Localhost 8546 alternative',
+                nativeCurrency: {
+                  name: '',
+                  symbol: 'ETH',
+                  decimals: 18,
+                },
+                // this does not match what already exists in the NetworkController
+                rpcUrls: ['http://127.0.0.1:8546'],
+                blockExplorerUrls: [],
+              },
+            ],
+          });
+
+          await driver.executeScript(
+            `window.ethereum.request(${switchEthereumChainRequest})`,
+          );
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+          await driver.clickElement({ text: 'Approve', tag: 'button' });
+
+          await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+          const afterPermittedChains = await getPermittedChains(driver);
+          assert.deepEqual(afterPermittedChains, ['0x539', '0x53a']);
+
+          // should end on 1338
+          await driver.findElement({ css: '#chainId', text: '0x53a' });
+        },
+      );
+    });
+
+    it('automatically switches to the chain when the rpc endpoint being added already exists for the chain', async function () {
```

Comment:
Will just highlight the case that (I think) we're missing here, which I've manually tested is, if the rpcEndpoint exists but is not currently the selected rpcEndpoint,, but the chain is selected

Reply:
agreed

Reply:
Add missing test case here: https://github.com/MetaMask/metamask-extension/pull/29885